apps/web/vibes/soul/primitives/navigation/index.tsx

```diff
@@ -507,19 +243,361 @@ export const Navigation = forwardRef(function Navigation(
+      </div>
+    </NavigationMenu.Root>
+  )
+})
+
+function HamburgerMenuButton({
+  open,
+  className,
+  ...rest
+}: { open: boolean } & React.ComponentPropsWithoutRef<'button'>) {
+  return (
+    <button
+      {...rest}
+      className={clsx(
+        'group relative rounded-lg p-2 outline-0 ring-primary transition-colors focus-visible:ring-2',
+        className
+      )}
+    >
+      <div className="flex h-4 w-4 origin-center transform flex-col justify-between overflow-hidden transition-all duration-300">
+        <div
+          className={clsx(
+            'h-px origin-left transform bg-foreground transition-all duration-300',
+            open ? 'translate-x-10' : 'w-7'
+          )}
+        />
+        <div
+          className={clsx(
+            'h-px transform rounded bg-foreground transition-all delay-75 duration-300',
+            open ? 'translate-x-10' : 'w-7'
+          )}
+        />
+        <div
+          className={clsx(
+            'h-px origin-left transform bg-foreground transition-all delay-150 duration-300',
+            open ? 'translate-x-10' : 'w-7'
+          )}
+        />
+
+        <div
+          className={clsx(
+            'absolute top-2 flex transform items-center justify-between bg-foreground transition-all duration-500',
+            open ? 'w-12 translate-x-0' : 'w-0 -translate-x-10'
+          )}
+        >
+          <div
+            className={clsx(
+              'absolute h-px w-4 transform bg-foreground transition-all delay-300 duration-500 ',
+              open ? 'rotate-45' : 'rotate-0'
+            )}
+          />
+          <div
+            className={clsx(
+              'absolute h-px w-4 transform bg-foreground transition-all delay-300 duration-500',
+              open ? '-rotate-45' : 'rotate-0'
+            )}
+          />
+        </div>
+      </div>
+    </button>
+  )
+}
+
+function SearchForm({
+  searchAction,
+  searchParamName = 'query',
+  searchHref = '/search',
+  searchInputPlaceholder = 'Search Products',
+  searchCtaLabel = 'View more',
+  emptySearchTitle,
+  emptySearchSubtitle,
+}: {
+  searchAction: SearchAction
+  searchParamName?: string
+  searchHref?: string
+  searchCtaLabel?: string
+  searchInputPlaceholder?: string
+  emptySearchTitle?: string
+  emptySearchSubtitle?: string
+}) {
+  const [query, setQuery] = useState('')
+  const [debouncedQuery, setDebouncedQuery] = useState('')
+  const debouncedOnChange = useMemo(() => debounce(setDebouncedQuery, 300), [])
+  const [{ searchResults, lastResult }, formAction, isPending] = useActionState(searchAction, {
+    searchResults: null,
+    lastResult: null,
+  })
+
+  useEffect(() => {
+    if (lastResult?.error) console.log(lastResult.error)
+  }, [lastResult])
+
+  useEffect(() => {
+    if (debouncedQuery === '') return
+
+    startTransition(() => {
+      const formData = new FormData()
+      formData.append(searchParamName, debouncedQuery)
+
+      formAction(formData)
```

Comment:
I still don't love that we're using actions for search but I don't have a better alternative yet...